### DISCUSSION TITLE
Doesn’t Ollama already support OpenAI api?

### DISCUSSION THREAD

**Discussion by NeverOccurs** [microsoft/autogen#2107] [General]
Then why the need to wrap up Ollama models with LiteLLM?

**Comment by ekzhu** [comment]
@marklysze would you like to elaborate?

**Comment by marklysze** [comment]
Hey @NeverOccurs, yep, Ollama does have an OpenAI-compatible API that can be used directly with AutoGen. If you don't need Function/Tool Calling, then I'd say Ollama's API is the way to go (saves having to have LiteLLM).

LiteLLM supports OpenAI's Function/Tool calling API parts and, together with Ollama, you can use function calling with AutoGen.

We are looking to add a direct Ollama API example as well in the documentation.

**Reply by NeverOccurs** [reply]
Thanks, really appreciate it. Actually I've been thinking about it a few days ago that if local models support function calling, then I saw this release. Will dig into it further.

**Comment by marklysze** [comment]
Anytime, if you want to chat with others interested in local models, feel free to join AutoGen's Discord - #alt-models channel is available for just that.

I'll close the discussion for now but feel free to create another here or on Discord if you have any.

**Comment by shuaib7860** [comment]
Hi sorry to open this issue again,

But does this also mean trying to use the RetreiveUserProxyAgent is not suitable for llama models just yet either because we need to do some function calling when trying to do a RAG agent?

**Reply by marklysze** [reply]
Hey @shuaib7860, if you're referring to Llama models (from Meta) that's different to Ollama which is used to run local models (including Llama as well as many more).

If you want to use Meta's Llama models (e.g. Llama 3) locally on your machine with AutoGen you can use Ollama however it isn't the best experience if you want to do both function calling and normal text generation. However, to improve the situation there's work being done on an Ollama specific client [see here](https://github.com/microsoft/autogen/pull/3056) that will allow you to use any local model through Ollama's Python library in AutoGen and it will support both normal text generation and function calling.

**Reply by shuaib7860** [reply]
Hey @marklysze, Thanks for getting back to me. Yes I was referring to using Llama 3 locally using Ollama to create a RAG agent using the RetreiveUserProxyAgent function and I keep getting this strange message of 

'Model llama3:70b not found. Using cl100k_base encoding.'

in the output. But the weird thing is that the tool does work and the agent does produce useful output and looking at my server the GPU is firing. So I'm quite confused about what is happening. I think it may be down to the fact that you say it currently isn't the best experience when combining autogen with Ollama. But not sure and wanted to confirm.

**Reply by marklysze** [reply]
Okay... If you're able to try using the Ollama PR code I linked to, it would be good to know if that works. That should be better at handling a mix of tool calling and normal text generation.

**Reply by shuaib7860** [reply]
Hey @marklysze, Apologies for the delay but I was travelling over the weekend. 

I gave your Olama PR code example a try with the Jack and Emma ConversableAgents and I get the same warning message, although it does work. 

![Screenshot 2024-07-15 at 13 33 16](https://github.com/user-attachments/assets/18fd5bcd-1fbc-4989-9fb6-1e0bc1e87fa2)

I attach a screenshot just for your reference.  Is anyone else seeing this warning too?

**Reply by shuaib7860** [reply]
Actually I have just thought of a potential reason why this warning message pops up but I'm not sure. So I'll share my thoughts and look forward to hearing yours in response. 

The LLM is hosted on a remote server and therefore I am port forwarding to the remote server from my local machine to interact with the LLM. Could this be the reason why this particular warning comes up as it looks to run the LLM locally and then just interacts with the one on the remote machine once the port forwarding kicks in? 

Also one more thing to add is with the RetrieveUserProxyAgent tool the warning says 

"Model llama3:70b not found. Using cl100k_base encoding."

So a little different but in essence the same warning. But this warning can be traced back to the token_count_utils python file.

https://github.com/microsoft/autogen/blob/5b1dc3bf63b26ea4c9598901dc0095e345ce79b4/autogen/token_count_utils.py#L82

**Reply by marklysze** [reply]
Hey @shuaib7860, sorry I haven't been that available over the last few weeks.

If you're referring to the message "...The cost will be 0", this is related to the way that AutoGen calculates inference costs for you. You can add that `price` field to your your config to set the price as zero. See [this comment](https://github.com/microsoft/autogen/issues/2555#issuecomment-2208623783) for an example of how to do that.

For the base encoding warning, and thanks for the reference to the token_count_utils file, this is just so that tokens can be counted. If you're not concerned by token costs then this can be ignored. I'm not sure if there's a way to turn it off.

**Reply by shuaib7860** [reply]
Many thanks @marklysze for the price field suggestion. And no worries on the time it took to get back to me. Thank you for all your help.
